nft_ingester/src/index_synchronizer.rs

```diff
@@ -68,39 +69,37 @@ where
 
     pub async fn nft_run(
         &self,
-        rx: &tokio::sync::broadcast::Receiver<()>,
+        cancellation_token: CancellationToken,
         run_full_sync_threshold: i64,
-        timeout_duration: tokio::time::Duration,
     ) {
-        while rx.is_empty() {
-            if let Err(e) = self.synchronize_nft_asset_indexes(rx, run_full_sync_threshold).await {
+        while !cancellation_token.is_cancelled() {
+            let cancellation_token = cancellation_token.child_token();
+            if let Err(e) = self
+                .synchronize_nft_asset_indexes(cancellation_token, run_full_sync_threshold)
+                .await
+            {
                 tracing::error!("Non fungible synchronization failed: {:?}", e);
             } else {
                 tracing::info!("Non fungible synchronization finished successfully");
             }
-            if rx.is_empty() {
-                tokio::time::sleep(timeout_duration).await;
-            }
         }
     }
 
     pub async fn fungible_run(
         &self,
-        rx: &tokio::sync::broadcast::Receiver<()>,
+        cancellation_token: CancellationToken,
         run_full_sync_threshold: i64,
-        timeout_duration: tokio::time::Duration,
     ) {
-        while rx.is_empty() {
-            if let Err(e) =
-                self.synchronize_fungible_asset_indexes(rx, run_full_sync_threshold).await
+        while !cancellation_token.is_cancelled() {
```

Comment:
The same concern as above

Reply:
each part in the subsequent call uses loops. the outer usage of cancellation token means that no new iterations will be started on this level, while the cancellation token in child functions acts to break inner loops